File: 144. Binary Tree Preorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    // recursive: 0ms, O(n)
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> traversal;
        if(!root) return traversal;
        helper(root, traversal);
        return traversal;
    }

    void helper(TreeNode* root, vector<int>& traversal){
        if(!root) return;
        traversal.push_back(root->val);
        helper(root->left, traversal);
        helper(root->right, traversal);
    }
};

File: 144. Binary Tree Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    
    # recursive dfs: O(N)time, O(h) space
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        def dfs(root, ret):
            if root:
                ret.append(root.val)
                dfs(root.left, ret)
                dfs(root.right, ret)
        ret = []
        dfs(root, ret)
        return ret
    
    # iterative, stack: O(N)time, O(N) space
    def preorderTraversal(self, root):
        ret, stack = [], [root]
        while stack:
            node = stack.pop()
            if node:
                ret.append(node.val)
                stack.append(node.right)
                stack.append(node.left)
        return ret
        